onnxruntime/core/graph/graph.cc

```diff
@@ -4046,6 +4070,210 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
+  const auto graph_output_defs = graph_to_inline.GetOutputs();
+  ORT_ENFORCE(node_output_defs.size() >= graph_output_defs.size());
+  for (size_t i = 0; i < graph_output_defs.size(); ++i) {
+    name_mapping.emplace(graph_output_defs[i]->Name(), node_output_defs[i]->Name());
+  }
+
+  // Process constant nodes and initializers first
+  // to create defs from them to retain the type, and make renaming easier
+  for (const auto& node : graph_to_inline.Nodes()) {
+    if (node.OpType() == kConstant) {
+      // Copy constant nodes _value to name_to_initial_tensor_
+      ONNX_NAMESPACE::NodeProto constant_node_proto{};
+      node.ToProto(constant_node_proto);
+
+      const auto& constant_name = constant_node_proto.output(0);
+      // Check if this is an output of the graph
+      std::string new_name;
+      auto hit = name_mapping.find(constant_name);
+      if (hit != name_mapping.cend()) {
+        new_name = hit->second;
+      } else {
+        new_name = GenerateNodeArgName(make_unique(constant_name));
+        name_mapping.emplace(constant_name, new_name);
+      }
+      ORT_RETURN_IF_ERROR(AddConstantProtoAsInitializer(constant_node_proto, new_name));
+    }
+  }
+
+  for (const auto& init : graph_to_inline.name_to_initial_tensor_) {
+    const gsl::not_null<TensorProto*> tensor{graph_proto_->add_initializer()};
+    *tensor = *init.second;
+
+    const auto& init_name = init.second->name();
+    // Check if this is an output of the graph
+    auto hit = name_mapping.find(init_name);
+    if (hit != name_mapping.cend()) {
+      tensor->set_name(hit->second);
+    } else {
+      auto new_name = GenerateNodeArgName(make_unique(init_name));
+      ORT_IGNORE_RETURN_VALUE(name_mapping.emplace(init_name, new_name));
+      tensor->set_name(std::move(new_name));
+    }
+
+    auto insert_result = name_to_initial_tensor_.emplace(tensor->name(), tensor);
+    ORT_ENFORCE(insert_result.second, "Initializer name: ", tensor->name(), " from graph: ",
+                graph_to_inline.Name(), " conflicts with graph initializer. Check Specializing code.");
+
+    const NodeArg* node_arg = graph_to_inline.GetNodeArg(init_name);
+    ORT_ENFORCE(node_arg != nullptr,
+                "The graph being inlined is expected to have NodeArg for initializer: ", init.second->name());
+    ORT_IGNORE_RETURN_VALUE(GetOrCreateNodeArg(tensor->name(), node_arg->TypeAsProto()));
+
+#if !defined(DISABLE_SPARSE_TENSORS)
+    if (graph_to_inline.IsSparseInitializer(init.first)) {
+      ORT_IGNORE_RETURN_VALUE(sparse_tensor_names_.emplace(tensor->name()));
+    }
+#endif
+  }
+
+  // We want to make sure we get nodes in topological order
+  // because Constant folding may cause the nodes appear in
+  // a different order.
+  GraphViewer graph(graph_to_inline);
+  for (const auto node_idx : graph.GetNodesInTopologicalOrder()) {
+    const auto* node = graph_to_inline.GetNode(node_idx);
+    if (node == nullptr || node->OpType() == kConstant) {
+      continue;
+    }
```

Comment:
wouldn't have expected either of these to be possible. IIRC the graph viewer filters out removed nodes, and any Constant was converted to an initializer in the ctor of the underlying Graph instance.